### PR TITLE
[SYCL][CUDA][HIP] Add piProgramHasKernel to PI

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This patch adds a function piProgramHasKernel to opencl, CUDA, and HIP. It removes CUDA and HIP's implementation of `piProgramGetInfo` with case `PI_PROGRAM_INFO_KERNEL_NAMES`. This is because the method was failing with per-source module splitting. The failing implementation used a regex search to try and find all the kernel names. 

As 'PI_PROGRAM_INFO_KERNEL_NAMES' is only used in `program_impl::has_kernel` a more robust implementation using `cu/hipModuleGetFunction` is proposed. This requires moving `has_kernel` behaviour inside of the PI.

Other smaller changes are made for module splitting per-source to work.

This patch does change the ABI but does not break it.

### CODE REVIEW

File: sycl/plugins/opencl/pi_opencl.cpp [hunk 1]
```diff
@@ -657,6 +657,60 @@ pi_result piProgramCreateWithBinary(
...
+  if (Res != CL_SUCCESS) {
+    *has_kernel = false;
+    return cast<pi_result>(Res);
+  }
+
+  // Get rid of the null terminator
+  ClResult.pop_back();
+  std::vector<std::string> KernelNames(split_string(ClResult, ';'));
+  for (const auto &Name : KernelNames) {
+    if (Name == kernel_name) {
```
Comment (steffenlarsen) [inline sycl/plugins/opencl/pi_opencl.cpp:705]: It seems like we do more work here than we need to. First we split `ClResult` into its constituents and then we look for one specific name inside that list. It would make sense to change `split_string` into a function that checks for a given string between the delimiters, rather than saving every string.

File: sycl/source/detail/queue_impl.hpp [hunk 2]
```diff
@@ -447,10 +447,8 @@ class queue_impl {
       bool IsKernel = Type == CG::Kernel;
       bool KernelUsesAssert = false;
       if (IsKernel)
-        KernelUsesAssert =
-            Handler.MKernel ? true
-                            : ProgramManager::getInstance().kernelUsesAssert(
-                                  Handler.MOSModuleHandle, Handler.MKernelName);
```
Comment (steffenlarsen) [inline sycl/source/detail/queue_impl.hpp:453]: Preferably this and the changes in sycl/include/CL/sycl/queue.hpp to be a separate PR as they seem fairly unrelated to the primary issues addressed by this PR.

**Review by steffenlarsen — COMMENTED** [review]
Good work! I am very happy to see the lackluster `getKernelNames` implementation disappear. I have a few comments though.

File: sycl/plugins/opencl/pi_opencl.cpp (hunk 1 above)
Comment (AidanBeltonS) [reply sycl/plugins/opencl/pi_opencl.cpp:705]: Sounds good. The implementation comes from the old `has_cl_kernel` function. I agree that the approach should be changed. Ill update the function.

File: sycl/source/detail/queue_impl.hpp (hunk 2 above)
Comment (AidanBeltonS) [reply sycl/source/detail/queue_impl.hpp:453]: Yes, this is to get module splitting per source to work, which is partly why I am making this PR. I agree that considering what I have written in the description this is not in scope. I'll remove sections related to this and make another PR.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 3]
```diff
@@ -567,30 +567,6 @@ pi_result _pi_program::build_program(const char *build_options) {
...
 }
 
-/// Finds kernel names by searching for entry points in the PTX source, as the
-/// CUDA driver API doesn't expose an operation for this.
-/// Note: This is currently only being used by the SYCL program class for the
-///       has_kernel method, so an alternative would be to move the has_kernel
-///       query to PI and use cuModuleGetFunction to check for a kernel.
-/// Note: Another alternative is to add kernel names as metadata, like with
-///       reqd_work_group_size.
-std::string getKernelNames(pi_program program) {
```
Comment (kbobrovs) [inline sycl/plugins/cuda/pi_cuda.cpp:577]: maybe DIE_NOIMPLEMENTATION?
Listing kernels will likely be needed for dynamic device code linkage. @AlexeySachkov - FYI

**Comment by smaslov-intel** [comment]
Can't we use existing `piextGetDeviceFunctionPointer` interface and check if the pointer returned is non-null?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -3528,6 +3528,58 @@ pi_result piProgramGetInfo(pi_program Program, pi_program_info ParamName,
...
+  ret_err = piProgramGetInfo(program, PI_PROGRAM_INFO_KERNEL_NAMES,
+                             ClResult.size(), &ClResult[0], nullptr);
+  if (ret_err != PI_SUCCESS) {
+    *has_kernel = false;
+    return ret_err;
+  }
+
+  // Get rid of the null terminator and search for kernel_name
+  ClResult.pop_back();
+  *has_kernel = is_in_seperated_string(ClResult, ';', (std::string)kernel_name);
```
Comment (steffenlarsen) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3579]: Cast to `std::string` shouldn't be necessary.

**Comment by steffenlarsen** [comment]
> Can't we use existing `piextGetDeviceFunctionPointer` interface and check if the pointer returned is non-null?

Implementation of `piextGetDeviceFunctionPointer` in sycl/plugins/opencl/pi_opencl.cpp contains a comment saying `TODO: add check that device supports corresponding extension`. Having `piProgramHasKernel` implemented like suggested here may be the more stable solution for now.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 3 above)
Comment (AidanBeltonS) [reply sycl/plugins/cuda/pi_cuda.cpp:577]: Good point. I'll make the change.

**Comment by smaslov-intel** [comment]
> > Can't we use existing `piextGetDeviceFunctionPointer` interface and check if the pointer returned is non-null?
> 
> Implementation of `piextGetDeviceFunctionPointer` in sycl/plugins/opencl/pi_opencl.cpp contains a comment saying `TODO: add check that device supports corresponding extension`. Having `piProgramHasKernel` implemented like suggested here may be the more stable solution for now.

Today there is no check for the `clGetDeviceFunctionPointerINTEL` extension, which suggests it is always there. And if it there I think it is indeed more robust to use it rather than the search you are adding. To be extra safe you could do your search, if extension is not available, and return fake "address-is-unknown" (non-null) pointer to represent that the kernel is there, but we can't tell its address.

It sounds better to re-use the existing very similar API than adding new one, which would stick.

**Review by kbobrovs — APPROVED** [review]
LGTM. But approval from @smaslov-intel is needed.

**Comment by steffenlarsen** [comment]
> Today there is no check for the `clGetDeviceFunctionPointerINTEL` extension, which suggests it is always there. And if it there I think it is indeed more robust to use it rather than the search you are adding. To be extra safe you could do your search, if extension is not available, and return fake "address-is-unknown" (non-null) pointer to represent that the kernel is there, but we can't tell its address.
> It sounds better to re-use the existing very similar API than adding new one, which would stick.

The solution suggested here is mostly just moved from the runtime, so I would go as far as to say it is proven more reliable than the use of `piextGetDeviceFunctionPointer`, as the latter is only used in a SYCL extension and is only implemented for some backends. That said, I agree that it would be preferable to use existing APIs. I also believe that it should be fairly straight forward to implement `piextGetDeviceFunctionPointer` for the CUDA backend by using [`cuModuleGetFunction`](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MODULE.html#group__CUDA__MODULE_1ga52be009b0d4045811b30c965e1cb2cf). However, I cannot help but wonder if this comes with an additional overhead.

The fallback to the name-search seems like a good alternative. The CUDA backend should hopefully always be able to answer `piextGetDeviceFunctionPointer` so it should never fall into the old name-search approach.

**Comment by AidanBeltonS** [comment]
I agree that it would be nice to reuse an already existing function.
To clarify, the search function used in OpenCL and level_zero is modified from the previous implementation of `has_kernel`.

`piextGetDeviceFunctionPointer` finds the functions poionter for a device. However, Cuda modules are not tied to a device so cuModuleGetFunction would return a pointer that is device independent. Looking at the docs, Cuda does not appear to be a reasonable equivalent, so `piextGetDeviceFunctionPointer`'s implementation may not behave the same as OpenCL's.

From my understanding, correct me if I am wrong, some devices could have access to different kernels for a program.
As a result, the `has_kernel` function would have to loop through each device to find if the kernel is held by the program. This would be a pretty minor issue but might be a bit messy.
So using `piextGetDeviceFunctionPointer` would certainly be possible, but it may be the less clean option.

**Comment by steffenlarsen** [comment]
> `piextGetDeviceFunctionPointer` finds the functions poionter for a device. However, Cuda modules are not tied to a device so cuModuleGetFunction would return a pointer that is device independent. Looking at the docs, Cuda does not appear to be a reasonable equivalent, so `piextGetDeviceFunctionPointer`'s implementation may not behave the same as OpenCL's.

In the CUDA driver API a `CUmodule` is tied to a CUDA context which is in turn tied to a single device. As such you should still be able to use `cuModuleGetFunction` as suggested.

> From my understanding, correct me if I am wrong, some devices could have access to different kernels for a program.
> As a result, the `has_kernel` function would have to loop through each device to find if the kernel is held by the program. This would be a pretty minor issue but might be a bit messy.
> So using `piextGetDeviceFunctionPointer` would certainly be possible, but it may be the less clean option.

I am not sure that is true. I would expect the PI program to have the same kernels on all the devices, but I am not sure that is guaranteed either. If I am wrong, then it may be a bit messy, but it should not be worse than iterate over the devices, finding the first one to return a pointer from `piextGetDeviceFunctionPointer`. Luckily for the current state of the CUDA backend, you don't have to worry as it will only have a single device always.

**Comment by AidanBeltonS** [comment]
> In the CUDA driver API a `CUmodule` is tied to a CUDA context which is in turn tied to a single device. As such you should still be able to use `cuModuleGetFunction` as suggested.

Ahh, okay I stand corrected. Then I think



> I am not sure that is true. I would expect the PI program to have the same kernels on all the devices, but I am not sure that is guaranteed either. If I am wrong, then it may be a bit messy, but it should not be worse than iterate over the devices, finding the first one to return a pointer from `piextGetDeviceFunctionPointer`. Luckily for the current state of the CUDA backend, you don't have to worry as it will only have a single device always.

Yes, so although the CUDA backend is always a single device, opencl is not and the loop would have to be outside of the PI. So if there could be a situation in which not all devices had the same kernel then you would still have to loop regardless of the backend. If this is the case (not sure), it is a pretty small problem.

I now agree that it is worth implementing `piextGetDeviceFunctionPointer` and scrap `piProgramHasKernel` as my main concern was that `cuModuleGetFunction` was not linked to a device.
